Fix the performance issue that the for-loop before ExternallCall could not be parallelized.

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85056

Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

EikanWang committed:
Fix the performance issue that the for-loop before ExternallCall could not be parallelized.

[ghstack-poisoned]

EikanWang commented:
Since we add more fusions by leveraging `ExternalCall`, this PR is more critical to Conv post-op fusions.

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]

EikanWang commented:
@ZolotukhinM, @frank-wei , could you help to prioritize this PR review and landing? Otherwise, the post-op fusion could introduce performance regression.

frank-wei reviewed (commented):
Add unit test?

EikanWang commented:
> Add unit test?

Sorry for the late reply. @frank-wei , yes, will do.

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]

frank-wei commented:
Could you pls check the failed tests?

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]

EikanWang commented:
> Could you pls check the failed tests?

Done. Refined the UT.

frank-wei reviewed (approved):
Ask @ZolotukhinM  for more opinions.

EikanWang commented:
@bertmaher , any comments?

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -670,29 +670,40 @@ bool loopBoundsAllEqual(const std::vector<ForPtr>& loops) {
-    std::vector<ForPtr> loopsToFuse;
-    for (auto stmt : *block) {
-      auto loop = to<For>(stmt);
-      if (!loop) {
```
It's concerning to me that this check, which aborts loop fusion if a non-loop statement is encountered, has been removed in this diff.  Specifically I'm concerned about a situation where there is a memory dependence, e.g.:
```
for i : (0, M)
  b[i] = a[i]
b[0] = 12
for i : (0, M)
  c[i] = b[i]
```
If we naively fuse those two loops, we'll miss the non-loop update of `b[0]` in line 3.  Now, it's possible that loop fusion itself will prevent this case (does it reject fusion of non-adjacent loops?  I can't remember).

In any case, please either point me to some existing tests that demonstrate correctness in such a case, write new tests exercising that behavior, or restore a check refusing to fuse things with non-loops.

EikanWang replied:
This loop enumerates the blocks within a given block. If the current block is not for-statement, the fusion would be broken.  Take the code snippet as an example, `b[0] = 12` is not a for-statement, hence the two loops will not be fused.

EikanWang replied:
This PR targets optimizing the loop parallelism. The original logic would not parallelize the first and second loop if the buffer is not the output of a fusion group.

EikanWang replied:
@bertmaher , this [UT](https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/test_loopnest.cpp#L5613) is to test the case. But I polished it a little bit by changing the loop interval to make sure the two loop intervals are the same but only a store is inserted between the two loops.

https://github.com/pytorch/pytorch/pull/85056/files#diff-27612a2830a19cb267524fa9e43a085d1c2de71a987754034da953c1c16b99e6R5628

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]

bertmaher reviewed (approved):
LGTM, thanks for the updated test case.

EikanWang committed:
Update on "Fix the performance issue that the for-loop before ExternallCall could not be parallelized."


Currently, NNC only parallelizes the loop statement of the graph outputs. The logic could bypass some loop statements that could be parallelized. Take an example as follows and suppose the output of `ExternallCall` is also the output of NNC fusion group. Current [parallel logic](https://github.com/pytorch/pytorch/pull/85056/files#diff-9a11174c26e4b57ab73e819520122bc314467c72962f3a5b79e7400ea3c4bbe5L781-L785) only tries to parallel the `ExternalCall` and bypass `stmt1` and `stmt2`.

```c++
stmt1: For:
stmt2:   For:
stmt3: ExternalCall
```

[ghstack-poisoned]
